### PR TITLE
arch/riscv: update ISR code to account for RISC-V cpus without 'mul'

### PR DESCRIPTION
**Pull request by nathankrueger**
This change adds a loop-based alternative for acquiring the CPU struct
in the Zephyr kernel associated with the currently executing HART.
In order to maintain compatibility with all currently defined SoC,
default the options for HW mul / Atomics support to 'y', but allow
them to be overridden for any SoC which does not support these

I tested this change locally via twister agaisnt a few RISC-V platforms
including some 32bit and 64bit. To verify the 4 possibilities of Atomics
& HW Mul: (No, No), (No, Yes), (Yes, No), (Yes, Yes -- current behavior),
I used an out-of-tree GCC (xPack RISC-V GCC) which has multilib support
for rv32i, rv32ia, rv32ima to test against our out-of-tree Intel Nios V/m
processor in HW.  The Zephyr SDK RISCV GCC currently does not contain
multilib support for all variants exposed by these new KConfig options.

Signed-off-by: Nathan Krueger <nathan.krueger@intel.com>

### CODE REVIEW

**Review by dnltz — COMMENTED** [review]
Awesome! Just some points from my side.

**Review by npitre — CHANGES_REQUESTED** [review]
Please split compiler ISA flags and cur_cpu computation in 2 separate commits.

File: arch/riscv/Kconfig [hunk 1]
```diff
@@ -13,6 +13,14 @@ config COMPRESSED_ISA
 	bool
 	default y if 64BIT
 
+config RISCV_ATOMICS_ISA
```
[the remark this answers is not in the document]
Comment (nathankrueger) [reply arch/riscv/Kconfig:16]: Shall I append the RISCV_ prefix to it?  This means I need to change an extra 7 files my by count.

Alternatively I can remove the RISCV_ prefix from my variables -- I had chosen it as I'm not sure of the namespace of KConfig symbols and don't want to clash with anything else.

File: arch/riscv/core/isr.S [hunk 2]
```diff
@@ -215,12 +215,34 @@
...
+/*
+ * @brief Return the cpu struct associated with the current
+ *        HART by indexing into the global kernel struct
+ *
+ * @param ret Register to return the value
+ * @param temp A temporary register you can safely clobber
+ *
+ * @return addr(_kernel) + (__cpu_t_SIZEOF * m_hartid)
+ */
+#ifdef CONFIG_RISCV_MUL_ISA
```
[the remark this answers is not in the document]
Comment (nathankrueger) [reply arch/riscv/core/isr.S:227]: I like it, ISR latency is an important thing for RTOS.  I will do it in a separate commit per @npitre's request.

File: arch/riscv/Kconfig (hunk 1 above)
[the remark this answers is not in the document]
Comment (dnltz) [reply arch/riscv/Kconfig:16]: Looks like other architectures tend to add the prefix. If it's not too much effort, why not?

**Comment by stephanosio** [comment]
> > The Zephyr SDK RISCV GCC currently does not contain
> > multilib support for all variants exposed by these new KConfig options.
> 
> Is this something we should look into @stephanosio?

@carlescufi Currently, Zephyr SDK supports the following multilibs for RISC-V:

```
rv32im/ilp32;@march=rv32im@mabi=ilp32
rv32imac/ilp32;@march=rv32imac@mabi=ilp32
rv32imafc/ilp32f;@march=rv32imafc@mabi=ilp32f
rv32imafd/ilp32d;@march=rv32imafd@mabi=ilp32d
rv32em/ilp32e;@march=rv32em@mabi=ilp32e
rv32emc/ilp32e;@march=rv32emc@mabi=ilp32e
rv64imac/lp64;@march=rv64imac@mabi=lp64
rv64imac/lp64/medany;@march=rv64imac@mabi=lp64@mcmodel=medany
rv64imafdc/lp64d;@march=rv64imafdc@mabi=lp64d
rv64imafdc/lp64d/medany;@march=rv64imafdc@mabi=lp64d@mcmodel=medany
rv64imafd/lp64d;@march=rv64imafd@mabi=lp64d
rv64imafd/lp64d/medany;@march=rv64imafd@mabi=lp64d@mcmodel=medany
```

What is missing, in terms of this PR, is:

* rv32i
* rv32ia
* rv32ima
* rv64i
* rv64ia
* rv64im
* rv64ima

We can look into adding the multilibs for these ISA extension combinations to the Zephyr SDK, but the question is:

1. Is anyone actually using these extension combinations?
1. Are these extension combinations desirable and/or useful? Do they make sense for any practical applications?
    (i.e. is anyone going to build a SoC with a RISC-V core configured as such?)

Technically, it is possible to add the multilibs for every possible extension combination; but, that is not going to scale well for obvious reasons, and we need to draw a line somewhere.

I would like to get some feedback on what extension combinations to add to the Zephyr SDK from the RISC-V maintainers.

File: arch/riscv/Kconfig (hunk 1 above)
Comment (stephanosio) [inline arch/riscv/Kconfig:16]: All these ISA-related configs will need to be refactored at some point, to be something like:

```
choice
	...
	default RISCV_ISA_RV64I if 64BIT
	default RISCV_ISA_RV32I

config RISCV_ISA_RV32I
	...

config RISCV_ISA_RV32E
	...

config RISCV_ISA_RV64I
	...

endchoice

config RISCV_ISA_EXT_M
	...

config RISCV_ISA_EXT_A
	...

config RISCV_ISA_EXT_G
	...
	imply RISCV_ISA_EXT_M
	imply RISCV_ISA_EXT_A
	imply RISCV_ISA_EXT_F
	imply RISCV_ISA_EXT_D
```

[the remark this answers is not in the document]
Comment (nathankrueger) [reply arch/riscv/Kconfig:16]: Should this be a separate PR?  I can add another commit if no.

File: arch/riscv/core/isr.S [hunk 3]
```diff
@@ -215,12 +215,34 @@
...
+#define GET_CPU(ret, temp1, temp2) \
+	csrr ret, mhartid			;\
+	li temp1, ___cpu_t_SIZEOF	;\
+	mul temp1, temp1, ret		;\
+	la ret, _kernel				;\
+	add ret, ret, temp1			;
```
[the remark this answers is not in the document]
Comment (nathankrueger) [reply arch/riscv/core/isr.S]: Fair enough, does make it more similar, addressed in my latest commit

File: arch/riscv/core/isr.S [hunk 4]
```diff
@@ -215,12 +215,34 @@
...
-	li temp, ___cpu_t_SIZEOF	;\
-	mul temp, temp, ret		;\
-	la ret, _kernel			;\
-	add ret, ret, temp		;
+/*
+ * @brief Return the cpu struct associated with the current
+ *        HART by indexing into the global kernel struct
+ *
+ * @param ret Register to return the value
+ * @param temp A temporary register you can safely clobber
```
[the remark this answers is not in the document]
Comment (nathankrueger) [reply arch/riscv/core/isr.S]: Good catch, addressed in my latest commit

File: arch/riscv/core/isr.S (hunk 2 above)
[the remark this answers is not in the document]
Comment (nathankrueger) [reply arch/riscv/core/isr.S:229]: Addressed in the latest commit

**Comment by nathankrueger** [comment]
> Please split compiler ISA flags and cur_cpu computation in 2 separate commits.

Done.

**Review by npitre — DISMISSED** [review]

**Review by edersondisouza — DISMISSED** [review]

**Review by npitre — CHANGES_REQUESTED** [review]
Please have a look at #43085. I used a better solution than a loop.

**Comment by nathankrueger** [comment]
> Please have a look at #43085. I used a better solution than a loop.

I looked and it does not pass twister -- I'd rather not debug it.

Can you rebase against my change which is fully functional and offers an improvement?

**Comment by npitre** [comment]
> I looked and it does not pass twister -- I'd rather not debug it.

It does pass now. I obviously didn't expect you to debug it.

> Can you rebase against my change which is fully functional and offers an
> improvement?

Your particular change might work properly, but the codebase you applied
your change to is far from being fully functional unfortunately. So I'd
encourage you to have another look. It'll be far easier for you to port
the remaining of your changes on top of that instead.

**Comment by nathankrueger** [comment]
> > I looked and it does not pass twister -- I'd rather not debug it.
> 
> It does pass now. I obviously didn't expect you to debug it.
> 
> > Can you rebase against my change which is fully functional and offers an
> > improvement?
> 
> Your particular change might work properly, but the codebase you applied your change to is far from being fully functional unfortunately. So I'd encourage you to have another look. It'll be far easier for you to port the remaining of your changes on top of that instead.



> > I looked and it does not pass twister -- I'd rather not debug it.
> 
> It does pass now. I obviously didn't expect you to debug it.
> 
> > Can you rebase against my change which is fully functional and offers an
> > improvement?
> 
> Your particular change might work properly, but the codebase you applied your change to is far from being fully functional unfortunately. So I'd encourage you to have another look. It'll be far easier for you to port the remaining of your changes on top of that instead.

Ok fair enough.

The 10x reduction of cycles is indeed pretty sweet.  Is there a particular part of your change that is primarily responsible for this?

**Comment by npitre** [comment]
> The 10x reduction of cycles is indeed pretty sweet.  Is there a particular part of your change that is primarily responsible for this?

Getting rid of all the necessary overhead in the exception code. Notably 
not going in and out then in and out again of exception mode for every 
syscall invocation, not copying the esf across different stacks, etc. 
There were many such abominations in that code.

**Comment by MaureenHelm** [comment]
#43085 has now been merged, please rebase

**Comment by nathankrueger** [comment]
> #43085 has now been merged, please rebase

Done!

**Review by MaureenHelm — APPROVED** [review]

**Review by npitre — APPROVED** [review]
Might be worth changing the title of this PR too.

**Review by nashif — APPROVED** [review]

**Comment by nashif** [comment]
@kgugala please take a look.

**Review by tgorochowik — APPROVED** [review]
LGTM, very useful for vexriscv

**Review by kgugala — APPROVED** [review]
LGTM

**Pull request merged**